### ISSUE TITLE
DateFormat parameter ignored [DATAES-287]

### ISSUE THREAD

**Issue by spring-projects-issues** [spring-projects/spring-data-elasticsearch#862]
**[Antoine Dupuis](https://jira.spring.io/secure/ViewProfile.jspa?name=mnshdw)** opened **[DATAES-287](https://jira.spring.io/browse/DATAES-287?redirect=false)** and commented

Hello guys,

When using spring-data-elasticsearch to save a Date field, the DateFormat parameter is never taken into account. This is because `DefaultEntityMapper.mapToString` uses a default Jackson ObjectMapper. As the result, dates are always saved as timestamps (long).

This was not a problem in Elasticsearch 1.x as date fields were always parsed first as timestamps. However in 2.x the format registered in the mapping is now properly used, so all dates are unparseable. See https://www.elastic.co/guide/en/elasticsearch/reference/current/breaking_20_mapping_changes.html#_date_fields_and_unix_timestamps

> Previously, date fields would first try to parse values as a Unix timestamp — milliseconds-since-the-epoch — before trying to use their defined date format. This meant that formats like yyyyMMdd could never work, as values would be interpreted as timestamps.
> In 2.0, we have added two formats: epoch_millis and epoch_second. Only date fields that use these formats will be able to parse timestamps.

If you don't have any data in your database yet, a possible workaround is to configure the Jackson mapper on the field directly. For example :

```
@Field(store = true, type = FieldType.Date, format = DateFormat.date_time)
@JsonFormat(shape=JsonFormat.Shape.STRING, pattern="yyyy-MM-dd'T'HH:mm:ss.SSSZZ")
private Date creationDate;
```

However, I would expect Spring to do that for me.

If you already have dates mapped with a date format but saved as a timestamp, you might get around the issue without reindexing all your data by updating the format in the mappings as suggested here : https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-date-format.html.

> The format setting must have the same setting for fields of the same name in the same index. Its value can be updated on existing fields using the PUT mapping API.

However, the new formats "epoch_millis" and "epoch_second" are currently not available in DateFomat !

TL;DR;
- Dates are always saved as timestamps (DateFormat not taken into account when serializing)
- Dates can't be deserialized from timestamps (lack of proper DateFormat)

How can I address this issue ?

Thanks !

---

**Affects:** 2.0.2 (Hopper SR2)

**Issue Links:**
- [DATAES-629](https://jira.spring.io/browse/DATAES-629) Move ElasticsearchEntityMapper to convert package and rename to MappingElasticsearchConverter
 (_**"is superseded by"**_)
- [DATAES-630](https://jira.spring.io/browse/DATAES-630) Remove GetResultMapper and friends from core package
 (_**"is superseded by"**_)

1 votes, 3 watchers
